packages/deployments/contracts/contracts/lib/Connext/ConnextLogic.sol

```diff
@@ -864,6 +878,177 @@ library ConnextLogic {
+
+    // Swap all the transfer amount from local asset to adopted if needed
+    (vars.adoptedAmountOut, vars.adopted) = AssetLogic.swapFromLocalAssetIfNeeded(
+      _canonicalToAdopted,
+      _adoptedToLocalPools,
+      _args.tokenRegistry,
+      _token,
+      _amount
+    );
+
+    // Calculates the amount to be repaid to the portal
+    (vars.totalRepayAmount, vars.backUnbackedAmount, vars.portalFee) = _calculatePortalRepayment(
+      vars.adoptedAmountOut,
+      vars.portalTransferAmount,
+      _args.portalFeeNumerator,
+      _args.portalFeeDenominator
+    );
+
+    // Edge case with some tokens: Example USDT in ETH Mainnet, after the backUnbacked call there could be a remaining allowance if not the whole amount is pulled by aave.
+    // Later, if we try to increase the allowance it will fail. USDT demands if allowance is not 0, it has to be set to 0 first.
+    // TODO: Should we call approve(0) and approve(totalRepayAmount) instead? or with a try catch to not affect gas on all cases?
+    // Example: https://github.com/aave/aave-v3-periphery/blob/ca184e5278bcbc10d28c3dbbc604041d7cfac50b/contracts/adapters/paraswap/ParaSwapRepayAdapter.sol#L138-L140
+    SafeERC20Upgradeable.safeIncreaseAllowance(IERC20Upgradeable(vars.adopted), _args.aavePool, vars.totalRepayAmount);
+
+    // TODO: What if backUnbacked call fails? For example if the asset is inactive, paused or frozen?
+    // What is the impact of a reconcile to be reverted? Can be re-executed later?
+    IAavePool(_args.aavePool).backUnbacked(vars.adopted, vars.backUnbackedAmount, vars.portalFee);
+
+    // TODO should emit some event with the repayment info?
+
+    // TODO: Do we need to check balance before and after to get the exact amount paid and give the routers the rest?
+    // Aave accounts a global unbacked variable per asset for all, not by address/bridge.
+    // Someone can repay more than it should, so then a the moment of calling backUnbacked()
+    // aave can pull a smaller amount than backUnbackedAmount. So there will be an extra amount of assets that needs to be assigned
+    // See https://github.com/aave/aave-v3-core/blob/feb3f20885c73025f40cc272b59e7eacfaa02fe4/contracts/protocol/libraries/logic/BridgeLogic.sol#L121
+
+    // Account leftovers in adopted to routers
+    return (vars.adoptedAmountOut - vars.totalRepayAmount, vars.adopted);
+  }
+
+  /**
+   * @notice Calculates the amount to be repaid to Aave Portal. If there is no enough amount to repay the unbacked and the fee,
+   * it will partially repay prioritizing the unbacked amount.
+   * @dev Assumes the fee is proportional to the unbackedAmount.
+   * @return The total amount to be repaid
+   * @return The unbacked amount to be backed
+   * @return The fee amount to be paid
+   */
+  function _calculatePortalRepayment(
+    uint256 availableAmount,
+    uint256 portalTransferAmount,
+    uint256 portalFeeNumerator,
+    uint256 portalFeeDenominator
```

Comment:
where are these assigned? ideally they should be pulling from the pool itself

Reply:
Answered in this other comment https://github.com/connext/nxtp/pull/1074#discussion_r863933180